File: client/scenes/Chunk.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;
using Godot;

namespace Data
{
	public class Chunk
	{
		public ulong id { get; set; }
		public string terrain_id { get; set; }
		public List<WorldObject> world_objects { get; set; }
		public List<Placable>? placables { get; set; }
		public static Chunk Deserialize(string json)
		{
			try
			{
				return JsonSerializer.Deserialize<Chunk>(json);
			}
			catch (Exception e)
			{
				GD.PrintErr(e.ToString());
			}
			return null;
		}

		public string GetModelUrl()
		{
			return "http://127.0.0.1:3000/assets/" + terrain_id;
		}
	}

	public interface IModelObject
	{
		public ulong id { get; set; }
		public string model_id { get; set; }
		public Vec3 position { get; set; }
		public Vec3 rotation { get; set; }

		public string GetModelUrl() { return "http://127.0.0.1:3000/assets/" + model_id; }
	}

	public class WorldObject : IModelObject
	{
		public ulong id { get; set; }
		public string model_id { get; set; }
		public Vec3 position { get; set; }
		public Vec3 rotation { get; set; }
		public List<Placable>? placables { get; set; }
	}

	public class Placable : IModelObject
	{
		public ulong id { get; set; }
		public string model_id { get; set; }
		public Vec3 position { get; set; }
		public Vec3 rotation { get; set; }
		public LightSource light_source { get; set; }
	}


	public class LightSource
	{
		public LightType light_type { get; set; }
		public Color color { get; set; }
		public float energy { get; set; }
		public float size { get; set; }
	}

	public enum LightType
	{
		SPOT,
		OMNI
	}
	public class Color
	{
		public float r { get; set; }
		public float g { get; set; }
		public float b { get; set; }

		public Godot.Color ToGodot()
		{
			return new Godot.Color(r, g, b);
		}
	}
	public class Vec3
	{
		public Vec3()
		{
			x = 0.0f;
			y = 0.0f;
			z = 0.0f;
		}
		public Vec3(float x, float y, float z)
		{
			this.x = x;
			this.y = y;
			this.z = z;
		}
		public float x { get; set; }
		public float y { get; set; }
		public float z { get; set; }

		public Vector3 ToGodot()
		{
			return new Vector3(x, y, z);
		}

		override public string ToString()
		{
			return "(" + x + ", " + y + ", " + z + ")";
		}

		public static Vec3 FromGodot(Vector3 vec3)
		{
			return new Vec3(vec3.X, vec3.Y, vec3.Z);
		}
	}

}

File: asset_server/src/routes/chunk.rs
use crate::models::world_data::*;
use axum::{extract::Path, http::StatusCode, response::IntoResponse, Json};

/// GET -> /chunk/{id}
/// returns JSON of chunk data
pub async fn get_chunk(Path(id): Path<u64>) -> impl IntoResponse {
    (StatusCode::OK, Json(get_chunk_info(id)))
}

fn get_chunk_info(id: u64) -> Chunk<'static> {
    //TODO: Search DB for chunk and return it
    Chunk {
        id,
        terrain_id: "floor.glb",
        world_objects: Some(vec![WorldObject {
            id: 222,
            model_id: "house.glb",
            position: Vec3 {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            },
            rotation: Vec3 {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            },
            placables: None,
        }]),
        placables: Some(vec![
            Placable {
                id: 78768,
                light_source: None,
                model_id: Some("TwistedTree_1.glb"),
                position: Vec3 {
                    x: 10.0,
                    y: 0.0,
                    z: 7.0,
                },
                rotation: Vec3 {
                    x: 0.0,
                    y: 0.0,
                    z: 0.0,
                },
            },
            Placable {
                id: 78768,
                light_source: None,
                model_id: Some("Plant_1_Big.glb"),
                position: Vec3 {
                    x: -2.2,
                    y: 0.0,
                    z: 4.0,
                },
                rotation: Vec3 {
                    x: 0.0,
                    y: 52.0,
                    z: 0.0,
                },
            },
            Placable {
                id: 78768,
                light_source: None,
                model_id: Some("Pine_1.glb"),
                position: Vec3 {
                    x: -11.0,
                    y: 0.0,
                    z: 6.5,
                },
                rotation: Vec3 {
                    x: 0.0,
                    y: 0.0,
                    z: 0.0,
                },
            },
            Placable {
                id: 78768,
                light_source: None,
                model_id: Some("Flower_3_Group.glb"),
                position: Vec3 {
                    x: 2.5,
                    y: 0.0,
                    z: 3.3,
                },
                rotation: Vec3 {
                    x: 0.0,
                    y: 0.0,
                    z: 0.0,
                },
            },
            Placable {
                id: 78768,
                light_source: Some(LightSource {
                    light_type: LightType::OMNI,
                    color: Color {
                        r: 1.0,
                        g: 0.0,
                        b: 0.0,
                    },
                    energy: 2.0,
                    size: 1.0,
                }),
                model_id: None,
                position: Vec3 {
                    x: 0.0,
                    y: 2.0,
                    z: 0.0,
                },
                rotation: Vec3 {
                    x: 0.0,
                    y: 0.0,
                    z: 0.0,
                },
            },
        ]),
    }
}
